In [13]:
# Netflix EDA Project
#Dataset: Netflix Movies & TV Shows (Kaggle)
#Tools: Python, Pandas, Matplotlib, Seaborn

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
print('Libraries loaded successfully!')



Libraries loaded successfully!


In [3]:
df = pd.read_csv('netflix_titles.csv')
print('Dataset loaded!')
print('Shape:', df.shape)


Dataset loaded!
Shape: (8807, 12)
Dataset loaded!
Shape: (8807, 12)


In [4]:
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [5]:
print('Column data types:')
print(df.dtypes)
print()
print('Missing values per column:')
print(df.isnull().sum())


Column data types:
show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

Missing values per column:
show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64
Column data types:
show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

Missing values per column:
show_id            0
type               0
title              

  ## Dataset Summary
  1. This dataset has 8807 rows and 12 columns.
  2. It contains Netflix movies and TV shows added between 2008 and 2021.
  3. Key columns include title, type, country, release_year, rating, and listed_in (genre).
  4. Columns like director and cast have the most missing values.
  5. The data will help us understand content patterns, growth trends, and audience targeting on Netflix.


In [6]:
print(df.isnull().sum())

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64


In [7]:
# Fill missing director with 'Unknown'
df['director'] = df['director'].fillna('Unknown')

# Fill missing cast with 'Not Available'
df['cast'] = df['cast'].fillna('Not Available')

# Fill missing country with 'Unknown'
df['country'] = df['country'].fillna('Unknown')

# Drop rows where date_added is missing (very few rows)
df = df.dropna(subset=['date_added'])

# Verify — all zeros means no missing values left
print('Missing values after cleaning:')
print(df.isnull().sum())



Missing values after cleaning:
show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          4
duration        3
listed_in       0
description     0
dtype: int64
Missing values after cleaning:
show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          4
duration        3
listed_in       0
description     0
dtype: int64


In [8]:
# Convert date_added from text to proper date format
df['date_added'] = pd.to_datetime(df['date_added'].str.strip())

# Extract just the year as a new column
df['year_added'] = df['date_added'].dt.year
print('date_added column type:', df['date_added'].dtype)
print('Sample year_added values:')
print(df['year_added'].value_counts().sort_index())


date_added column type: datetime64[ns]
Sample year_added values:
year_added
2008       2
2009       2
2010       1
2011      13
2012       3
2013      11
2014      24
2015      82
2016     429
2017    1188
2018    1649
2019    2016
2020    1879
2021    1498
Name: count, dtype: int64


  ## Data Cleaning Decisions
  - **director missing → filled with 'Unknown':** Most analysis does not need
    an exact director name. Dropping rows would lose too much data.
  - **cast missing → filled with 'Not Available':** Same reasoning as director.
  - **country missing → filled with 'Unknown':** Country is used in analysis,
    so we keep the rows but flag them as unknown.
  - **date_added missing → rows dropped:** Only a handful of rows affected.
    Without a date we cannot analyze time-based trends.
  - **date_added type fixed:** Converted from text string to datetime so we can
    extract year and month for trend analysis.


In [9]:
print('Final dataset shape:', df.shape)
df.head(3)


Final dataset shape: (8797, 13)
Final dataset shape: (8797, 13)


,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description,year_added
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Not Available,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021
1,s2,TV Show,Blood & Water,Unknown,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t...",2021
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",Unknown,2021-09-24,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...,2021


,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description,year_added
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Not Available,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021
1,s2,TV Show,Blood & Water,Unknown,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t...",2021
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",Unknown,2021-09-24,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...,2021


 ## EDA — Exploratory Data Analysis
  ### Q1: What types of content exist on Netflix?


In [10]:
  type_counts = df['type'].value_counts()
  print(type_counts)

  # Percentage breakdown
  print()
  print(type_counts / type_counts.sum() * 100)


type
Movie      6131
TV Show    2666
Name: count, dtype: int64

type
Movie      69.694214
TV Show    30.305786
Name: count, dtype: float64


**Finding:** Netflix has significantly more movies than TV shows.
  Movies make up roughly 70% of the library while TV shows account for 30%.
  This suggests Netflix historically licensed more films than it produced series.


### Q2: Which countries produce the most Netflix content?

In [11]:
  top_countries = df['country'].value_counts().head(10)
  print(top_countries)



country
United States     2812
India              972
Unknown            830
United Kingdom     418
Japan              244
South Korea        199
Canada             181
Spain              145
France             124
Mexico             110
Name: count, dtype: int64


  **Finding:** The United States dominates Netflix content by a large margin.
  India ranks second, reflecting Netflix's major push into the Indian market.
  South Korean content appears in the top 10, suggesting the K-drama boom is well represented.


  ### Q3: How has Netflix content volume changed over the years?

In [12]:
  yearly = df.groupby('year_added')['title'].count().reset_index()
  yearly.columns = ['year', 'count']
  print(yearly.sort_values('year').to_string(index=False))


 year  count
 2008      2
 2009      2
 2010      1
 2011     13
 2012      3
 2013     11
 2014     24
 2015     82
 2016    429
 2017   1188
 2018   1649
 2019   2016
 2020   1879
 2021   1498


  **Finding:** Netflix's content library grew slowly until 2015, then accelerated sharply.
  2019 was the peak year for content additions.
  There is a visible dip in 2020, likely due to production shutdowns during the COVID-19 pandemic.
